espnet2/asr/frontend/fused.py

```diff
@@ -0,0 +1,96 @@
+        self.frontends = []   # list of the frontends to combine
+
+        for i,frontend in enumerate(frontends) :
+            frontend_type = frontend["frontend_type"]
+            if frontend_type == "default":
+                n_mels, fs, n_fft, win_length, hop_length = frontend.get("n_mels", 80), fs, frontend.get("n_fft", 512), frontend.get("win_length"), frontend.get("hop_length", 128)
+                window, center , normalized , onesided = frontend.get("window", "hann"), frontend.get("center", True), frontend.get("normalized", False), frontend.get("onesided", True)
+                fmin, fmax, htk, apply_stft = frontend.get("fmin", None) , frontend.get("fmax", None) , frontend.get("htk", False) , frontend.get("apply_stft", True)
+
+                self.frontends.append(DefaultFrontend(n_mels=n_mels, n_fft=n_fft, fs=fs, win_length=win_length, hop_length=hop_length, window=window, \
+                                                      center=center, normalized=normalized, onesided=onesided, \
+                                                      fmin=fmin, fmax=fmax, htk=htk, apply_stft=apply_stft))
+            elif frontend_type == "s3prl":
+                frontend_conf, download_dir, multilayer_feature = frontend.get("frontend_conf"), frontend.get("download_dir"), frontend.get("multilayer_feature")
+                self.frontends.append(S3prlFrontend(fs=fs, frontend_conf=frontend_conf, download_dir=download_dir, multilayer_feature=multilayer_feature))
+
+            else :
+                raise NotImplementedError   # we only have those two categories of frontends for now
+
+        self.frontends = torch.nn.ModuleList(self.frontends)
+
+        self.gcd = np.gcd.reduce([frontend.hop_length for frontend in self.frontends])
+        self.factors = [frontend.hop_length // self.gcd for frontend in self.frontends]
+
+        if self.align_method == "linear_projection":
+            self.projection_layers = [torch.nn.Linear(in_features=frontend.output_size(),
+                                                          out_features=self.factors[i] * self.proj_dim, device="cuda") for i,frontend in enumerate(self.frontends)]
+            self.projection_layers = torch.nn.ModuleList(self.projection_layers)
+
+
+    def output_size(self) -> int:
+        return len(self.frontends) * self.proj_dim
+
+
+    def forward(
+            self, input: torch.Tensor, input_lengths: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+
+        # step 0 : get all frontends features
+        self.feats = []
+        for frontend in self.frontends :
+            with torch.no_grad():
+                input_feats, feats_lens = frontend.forward(input, input_lengths)
+            self.feats.append([input_feats, feats_lens])
+
+
+        if self.align_method == "linear_projection":   # for now, this is the only method provided for the fusion, others will be added in the next weeks
```

Comment:
For comment, please add a TODO here. The format could be

```
# TODO (Dan): to add other align methods
```

Reply:
ok